lib/streamlit/uploaded_file_manager.py

```diff
@@ -76,12 +69,13 @@ def add_files(self, session_id, widget_id, files):
             The files to add.
 
         """
-        file_list = UploadedFileList(
-            session_id=session_id, widget_id=widget_id, files=files
-        )
+        files_by_widget = session_id, widget_id
         with self._files_lock:
-            self._files_by_id[file_list.id] = file_list
-        self.on_files_added.send(file_list)
+            file_list = self._files_by_id.get(files_by_widget, None)
+            if file_list:
+                files = file_list + files
+            self._files_by_id[files_by_widget] = files
+        self.on_files_updated.send(files_by_widget)
```

Comment:
The logic here might change depending on product discussions. Primarily, discussions are around when we should rerun when file uploader has changes